parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/CompatibilityUtil.java

```diff
@@ -0,0 +1,396 @@
+    }
+  }
+  
+  static {
+    boolean v21 = true;
+    try {
+      Class.forName("org.apache.hadoop.io.ElasticByteBufferPool");
+    } catch (ClassNotFoundException cnfe) {
+      v21 = false;
+    }
+    
+    useV21 = v21;
+    try {
+      if (v21) {
+        fileAPI = new V21FileAPI();
+        bufferPool = fileAPI.ELASTIC_BYTE_BUFFER_CONSTRUCTOR.newInstance();
+      } else {
+        fileAPI = null;
+        bufferPool = null;
+      }
+    } catch (ClassNotFoundException e) {
+      throw new IllegalArgumentException("Can't find class", e);
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException("Can't find constructor ", e);
+    } catch (InstantiationException e) {
+      throw new IllegalArgumentException("Can't create instance ", e);
+    } catch (IllegalAccessException e) {
+      throw new IllegalArgumentException("Can't create instance ", e);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Can't create instance ", e);
+    } catch (InvocationTargetException e) {
+      throw new IllegalArgumentException("Can't create instance ", e);
+    }
+  }
+  
+  public static void releaseBuffer(FSDataInputStream f, ByteBuffer buf) {
+    if (useV21) {
+      try {
+        fileAPI.RELEASE_BUFFER_METHOD.invoke(f, buf);
+      } catch (IllegalAccessException e) {
+        throw new IllegalArgumentException("Can't call method", e);
+      } catch (IllegalArgumentException e) {
+        throw new IllegalArgumentException("Can't call method", e);
+      } catch (InvocationTargetException e) {
+        throw new IllegalArgumentException("Can't call method", e);
+      }
+     } 
+  }
+  
+  public static int getInt(FSDataInputStream f) throws IOException {
+    ByteBuffer int32Buf = getBuf(f, 4).order(ByteOrder.LITTLE_ENDIAN);
+    if (int32Buf.remaining() == 4) {
+      final int res = int32Buf.getInt();
+      releaseBuffer(f, int32Buf);
+      return res;
+    }
+    ByteBuffer tmpBuf = int32Buf;
+    int32Buf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
+    int32Buf.put(tmpBuf);
+    releaseBuffer(f, tmpBuf);
+    while (int32Buf.hasRemaining()) {
+      tmpBuf = getBuf(f, int32Buf.remaining());
+      int32Buf.put(tmpBuf);
+      releaseBuffer(f, tmpBuf);
+    }
+    return int32Buf.getInt();
+  }
+  
+  public static ByteBuffer getBuf(FSDataInputStream f, int maxSize)
+      throws IOException {
+    ByteBuffer res = null;
+    if (useV21) {
+      try {
+        res = (ByteBuffer) fileAPI.READ_METHOD.invoke(f,
+                                              fileAPI.ELASTIC_BYTE_BUFFER_CONSTRUCTOR.newInstance(),
+                                              maxSize,
+                                              EnumSet.of(Enum.valueOf(fileAPI.ReadOptionCls, "SKIP_CHECKSUMS")));
+      } catch (Exception e) {
+        byte[] buf = new byte[maxSize];
+        f.read(buf,0,  maxSize);
+        res = ByteBuffer.wrap(buf);
```

Comment:
to deal with the case where the FSDataInputStream does not actually support the new API, we should just wrap FSDataInputStream into our own class that encapsulates how we access it.
That way we can keep state for the access mode (ByteBuffer or Stream) and only catch the first exception.
We can have a static factory method:
```
static FSDataISAccessor newFSDataISAccessor(FSDataInputStream is)
```
where FSDataISAccessor has 2 strategies to access the data and can fall back when there's an exception.